### PR TITLE
Error codes for BsonP3

### PR DESCRIPTION
**Pull request by aaijazi**
@alainjobart 

Large clean-up of BsonP3, it now matches grpc almost exactly. As discussed, went the route of doing error codes serialized in the error string, so that we can keep our proto3 clean and take out the error  where it's not needed.

I'm doing a simple JSON serialization for the error string.

### CODE REVIEW

**Comment by aaijazi** [comment]
@enisoc I didn't realize that PHP was using BsonP3. What's the impact of changing the error reporting in this way?

File: go/vt/vtgate/bsonp3vtgateservice/server.go [hunk 1]
```diff
@@ -197,212 +186,141 @@ func (vtg *VTGateP3) ExecuteBatchKeyspaceIds(ctx context.Context, request *pb.Ex
...
 		return nil
 	}
-	return vtgErr
+	return vterrors.ToJSONError(vtgErr)
 }
 
-// Commit2 is the RPC version of vtgateservice.VTGateService method
-func (vtg *VTGateP3) Commit2(ctx context.Context, request *pb.CommitRequest, response *pb.CommitResponse) (err error) {
+// Commit is the RPC version of vtgateservice.VTGateService method
+func (vtg *VTGateP3) Commit(ctx context.Context, request *pb.CommitRequest, response *pb.CommitResponse) (err error) {
```
Comment (aaijazi) [inline go/vt/vtgate/bsonp3vtgateservice/server.go]: @alainjobart @enisoc 

We should decide how we want to name these, and be consistent about it.

gRPC uses Begin, Commit, Rollback, etc: https://github.com/youtube/vitess/blob/master/go/vt/vtgate/grpcvtgateservice/server.go#L306

BsonP3 was using Begin2, Commit2, Rollback2, etc. I changed things to be more consistent with gRPC, but that broke the PHP client.

I'm thinking it makes sense to name them <method>2 consistently, as that way we have a clearly versioned API (v1 doesn't support caller ID and has loose guarantees about error codes; v2 has full support for both).

That would require us to change gRPC, stubbyp3, etc to match. The legacy protocols already do what I'm describing above.

Things get a little unclear when it comes to Execute* though, as those didn't necessarily break compatibility... So there, we don't have clearly version APIs.

Thoughts?

Comment (aaijazi) [inline go/vt/vtgate/bsonp3vtgateservice/server.go]: Actually, thought about it some more: I'm in favor of everything based off of proto3 to just use the simple Begin, Commit, Rollback etc.

Everything that is not based off of p3 is slowly getting deprecated and therefore it's not a huge deal if they're not cleanly named.

I'm going to standardize on the non-2 version unless there are strong objections :)

**Comment by alainjobart** [comment]
I agree, go with the non-2 versions. That's the stable API names we want for release. Deprecate everything else. In google3, once we switch to the 2 or 3 versions, change the original versions to call the 2 or 3, and change the client to use that and remove the 2 or 3.

**Comment by alainjobart** [comment]
LGTM overall though, once versioning stuff is figured out. :)

**Comment by enisoc** [comment]
The PHP client has no tests yet for error handling, so I don't think the error changes should break it. I think all you should need to do is fix the method names since you removed the "2".

**Pull request merged**